### ISSUE TITLE
Uniqueness validation not triggering when calling create_association

### ISSUE THREAD

**Issue by rcpj** [rails/rails#48632]
From Rails version v7.0.5 onwards, the uniqueness validation stopped working when calling the create_association method from the model with the _has_one_. This doesn't happen if the create method is issued directly from the association model instead.

### Steps to reproduce
```ruby
#!/usr/bin/env ruby

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'rails', '~> 7.0.5'
  gem 'sqlite3'
  gem 'byebug'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :invoices do |t|
    t.string :name
  end

  create_table :payments do |t|
    t.string :name
    t.references :invoice
  end
end

class Invoice < ActiveRecord::Base
  has_one :payment, dependent: :destroy
end

class Payment < ActiveRecord::Base
  belongs_to :invoice, optional: true
  validates :invoice_id, uniqueness: true
end

class TestUniquenessValidationNotTriggering < Minitest::Test
  def test_uniqueness_validation_not_triggering
    assert_raises ActiveRecord::RecordInvalid do
      invoice = Invoice.create(name: 'Example invoice')
      invoice.create_payment!(name: "Alice's payment")
      assert_equal "Alice's payment", invoice.payment.name

      invoice.create_payment!(name: "Bob's payment")
    end
  end
end
```

### Expected behavior
It should raise ActiveRecord::RecordInvalid due to uniqueness validation

### Actual behavior
It creates the record successfully not raising the ActiveRecord::RecordInvalid exception

### System configuration
**Rails version**: ~> 7.0.5

**Ruby version**: 3.2.1

**Comment by fatkodima** [comment]
This is an expected change - https://github.com/rails/rails/pull/46386

**Comment by javierav** [comment]
> #46386

Can you explain why is expected? For me it`s very confused that the validation not apply in this case! And if it`s is expected... a breaking change released without entry in the changelog in a patch version of the framework may not be an ideal situation...

**Comment by pedroaugustofsilva** [comment]
Explaining a bit more of fatkondima comment.

When you call `invoice.create_payment!(name: "Bob's payment")` for the second time, as the `has_one` association has `dependent: :destroy`, it will destroy the old association and create a new one.

Before #46386, it would try to create the new record before deleting the older one, raising `RecordInvalid`.

**Comment by sinsoku** [comment]
Same issue as #48330

**Comment by akhilgkrishnan** [comment]
Seems like this issue is fixed #48808
